File: core/src/com/broll/gainea/server/core/actions/ActionHandlers.java
package com.broll.gainea.server.core.actions;

import com.broll.gainea.net.NT_Action_Card;
import com.broll.gainea.net.NT_Action_SelectChoice;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.actions.impl.AttackAction;
import com.broll.gainea.server.core.actions.impl.CardAction;
import com.broll.gainea.server.core.actions.impl.MoveUnitAction;
import com.broll.gainea.server.core.actions.impl.PlaceUnitAction;
import com.broll.gainea.net.NT_Action;
import com.broll.gainea.net.NT_Action_Attack;
import com.broll.gainea.net.NT_Action_Move;
import com.broll.gainea.net.NT_Action_PlaceUnit;
import com.broll.gainea.server.core.actions.impl.SelectChoiceAction;

import java.util.HashMap;
import java.util.Map;


public class ActionHandlers {

    private Map<Class<? extends NT_Action>, AbstractActionHandler> handlers = new HashMap<>();
    private Map<Class<? extends AbstractActionHandler>, AbstractActionHandler> handlers2 = new HashMap<>();

    private GameContainer game;
    private ReactionActions reactionResult;

    public ActionHandlers(GameContainer game, ReactionActions reactionResult) {
        this.game = game;
        this.reactionResult = reactionResult;
        initHandlers();
    }

    public ReactionActions getReactionActions() {
        return reactionResult;
    }

    private void initHandlers() {
        initHandler(NT_Action_PlaceUnit.class, new PlaceUnitAction());
        initHandler(NT_Action_Move.class, new MoveUnitAction());
        initHandler(NT_Action_Attack.class, new AttackAction());
        initHandler(NT_Action_Card.class, new CardAction());
        initHandler(NT_Action_SelectChoice.class, new SelectChoiceAction());
    }

    private void initHandler(Class<? extends NT_Action> actionClass, AbstractActionHandler handler) {
        handler.init(game, reactionResult, this);
        handlers.put(actionClass, handler);
        handlers2.put(handler.getClass(), handler);
    }

    public <T extends NT_Action> AbstractActionHandler<T, ?> getHandlerForAction(Class<T> actionClass) {
        return handlers.get(actionClass);
    }

    public <T extends AbstractActionHandler> T getHandler(Class<T> handlerClass) {
        return (T) handlers2.get(handlerClass);
    }

}

File: server/src/main/com/broll/gainea/server/core/actions/ActionHandlers.kt
package com.broll.gainea.server.core.actions

import com.broll.gainea.net.NT_Action
import com.broll.gainea.net.NT_Action_Attack
import com.broll.gainea.net.NT_Action_Card
import com.broll.gainea.net.NT_Action_Move
import com.broll.gainea.net.NT_Action_PlaceUnit
import com.broll.gainea.net.NT_Action_SelectChoice
import com.broll.gainea.net.NT_Action_Shop
import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.actions.optional.AttackAction
import com.broll.gainea.server.core.actions.optional.CardAction
import com.broll.gainea.server.core.actions.optional.MoveUnitAction
import com.broll.gainea.server.core.actions.optional.ShopAction
import com.broll.gainea.server.core.actions.required.PlaceUnitAction
import com.broll.gainea.server.core.actions.required.SelectChoiceAction

class ActionHandlers(private val game: Game, val reactionActions: ReactionActions) {
    private val handlers: MutableMap<Class<out NT_Action>, AbstractActionHandler<*, *>> = HashMap()
    private val handlers2: MutableMap<Class<out AbstractActionHandler<*, *>>, AbstractActionHandler<*, *>> =
        HashMap()

    init {
        initHandlers()
    }

    private fun initHandlers() {
        initHandler(NT_Action_PlaceUnit::class.java, PlaceUnitAction())
        initHandler(NT_Action_Move::class.java, MoveUnitAction())
        initHandler(NT_Action_Attack::class.java, AttackAction())
        initHandler(NT_Action_Card::class.java, CardAction())
        initHandler(NT_Action_SelectChoice::class.java, SelectChoiceAction())
        initHandler(NT_Action_Shop::class.java, ShopAction())
    }

    private fun initHandler(
        actionClass: Class<out NT_Action>,
        handler: AbstractActionHandler<*, *>
    ) {
        handler.init(game, reactionActions, this)
        handlers[actionClass] = handler
        handlers2[handler.javaClass] = handler
    }

    fun <T : NT_Action> getHandlerForAction(actionClass: Class<T>): AbstractActionHandler<T, ActionContext<out T>>? {
        return handlers.get(actionClass) as AbstractActionHandler<T, ActionContext<out T>>?
    }

    fun <T : AbstractActionHandler<*, *>> getHandler(handlerClass: Class<T>): T {
        return handlers2[handlerClass] as T
    }
}
